K0s docs tw extensions (#859)
k0sproject/k0s: docs/examples/traefik-ingress.md

# Install the Traefik Ingress Controller on k0s

You can configure k0s with the [Traefik ingress
controller](https://doc.traefik.io/traefik/providers/kubernetes-ingress/), a
[MetalLB service loadbalancer](https://metallb.universe.tf/),
and deploy the Traefik Dashboard using a service sample. With the extensible
bootstrapping functionality that Helm provides, this is as simple as adding the

> several extensions to the `k0s.yaml` file  when you configure your cluster.

Reviewer: "as simple as adding the several extensions" sounds wrong.

Author: Changed the sentence to read: "To do this you leverage Helm's extensible
bootstrapping functionality to add the correct extensions to the `k0s.yaml` file during cluster configuration."

## 1. Configure k0s.yaml

> Configure Traefik and MetalLB Helm charts to install during cluster bootstrap

Reviewer: maybe "to be installed"?

Or:

```suggestion
 "Configure k0s to install Traefik and MetalLB during cluster bootstrapping by adding their Helm charts as extensions in the k0s configuration file (`k0s.yaml`).
```
(and maybe somehow link to some other piece of documentation which tells what this extension business is about)

Author: Like the suggested change. Put a link in to the Helm charts topic.

by modifing your k0s configuration file (`k0s.yaml`) to include the Traefik and
MetalLB helm charts as extensions.

##### Note

A good practice is to have a small range of IP addresses for MetalLB on your

> LAN that are addressable, preferably outside the assignment pool your DHCP

Reviewer: The intent in the original snippet is "addressable on your network", this seems to change that to "IP addresses that are addressable", which, I think, does not compute.

Author: Tweaked it back for the most part. The note now reads: 

"A good practice is to have a small range of IP addresses  that are addressable on your network, preferably outside the assignment pool your DHCP server allocates (though any valid IP range should work locally on your machine). Providing an addressable range allows you to access your LoadBalancer and Ingress services from anywhere on your local network."

server allocates (though any valid IP range should work locally on your
machine). Providing an addressable range allows you to access your

> LoadBalancer and Ingress services from anywhere on your local network.

Reviewer: Seems there's some kind of mixture of "loadbalancer", "LoadBalancer" and "Load Balancer" in this doc. Maybe normalize to "load balancer", title-casing only when appearing in a title or product name.

            - 192.168.0.5-192.168.0.10
```

## 2. Retrieve the Load Balancer IP

After you start your cluster, run `kubectl get all` to confirm the deployment
of Traefik and MetalLB. The command should return a response with the `metallb`
and `traefik` resources, along with a service loadbalancer that has an assigned `EXTERNAL-IP`.

> ```bash
> root@k0s-host ➜ kubectl get all

Reviewer: ```suggestion
```shell
$ kubectl get all
```

Author: Fixed as per the $ Slack discussion.

404 page not found
```

## 3. Deploy and access the Traefik Dashboard

With an available and addressable load balancer present on your cluster, now
you can quickly deploy the Traefik dashboard and access it from anywhere on

> your LAN (assuming that MetalLB is configured with an addressable range).

Reviewer: If this is a tutorial, I think it should tell the user how to do it and assume the user did what was told, not suggest something and start questioning later if they did it or not.

Author: Removed the parenthetical remark, to friendly the sentence up and make the assumption that our reader is doing what they should be doing.

1. Create the Traefik Dashboard [IngressRoute](https://doc.traefik.io/traefik/providers/kubernetes-crd/) in a YAML file:

    ```yaml
    apiVersion: traefik.containo.us/v1alpha1
    kind: IngressRoute
    metadata:
      name: dashboard
    spec:
      entryPoints:
        - web
      routes:
        - match: PathPrefix(`/dashboard`) || PathPrefix(`/api`)
          kind: Rule
          services:
            - name: api@internal
              kind: TraefikService
    ```

2. Deploy the resource:

>     ```bash
>     root@k0s-host ➜ kubectl apply -f traefik-dashboard.yaml

Reviewer: ```suggestion
    ```shell
    $ kubectl apply -f traefik-dashboard.yaml
```

    ingressroute.traefik.containo.us/dashboard created
    ```

>     At this point you should be able to access the dashboard using the `EXTERNAL-IP` that you noted above by visiting `http://192.168.0.5` in your browser:

Reviewer: "that you noted above" => "from the previous step" 🤔 ?

Author: I changed this text to add a bit of future-proofing to the content. If ever. a step is added between creating the dashboard and deploying it...? Always better not to use references in procedure that are too HARD.

    ![Traefik Dashboard](../img/traefik-dashboard.png)

3. Create a simple `whoami` Deployment, Service, and [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) manifest:

    ```yaml
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: whoami-deployment
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: whoami
      template:
        metadata:
          labels:
            app: whoami
        spec:
          containers:
          - name: whoami-container
            image: containous/whoami
    ---
    apiVersion: v1
    kind: Service
    metadata:
      name: whoami-service
    spec:
      ports:
      - name: http
        targetPort: 80
        port: 80
      selector:
        app: whoami
    ---
    apiVersion: networking.k8s.io/v1
    kind: Ingress
    metadata:
      name: whoami-ingress
    spec:
      rules:
      - http:
          paths:
          - path: /whoami
            pathType: Exact
            backend:
              service:
                name: whoami-service
                port:
                  number: 80
    ```

4. Apply and test the manifest:

>     ```bash
>     # apply the manifests
>     root@k0s-host ➜ kubectl apply -f whoami.yaml

Reviewer: ```suggestion
    ```shell
    # apply the manifests
    $ kubectl apply -f whoami.yaml
```

    deployment.apps/whoami-deployment created
    service/whoami-service created
    ingress.networking.k8s.io/whoami-ingress created
    # test the ingress and service

>     root@k0s-host ➜ curl http://192.168.0.5/whoami

Reviewer: ```suggestion
    $ curl http://192.168.0.5/whoami
```
